File: methods/matlab/PCA.m
% @file PCA.m
% @author Marcus Edel
%
% Principal Components Analysis with matlab.

function pca(cmd)
% This program performs principal components analysis on the given dataset.
% It will transform the data onto its principal components, optionally 
% performing dimensionality reduction by ignoring the principal components 
% with the smallest eigenvalues.
%
% Required options:
%     (-i) [string]    Input dataset to perform PCA on.
% Options:
%     (-d) [int]       Desired dimensionality of output dataset. If this 
%                      option not set no dimensionality reduction is 
%                      performed. Default value 0.
%     (-s)             If set, the data will be scaled before running PCA,
%                      such that the variance of each feature is 1.


inputFile = regexp(cmd, '.*?-i ([^\s]+)', 'tokens', 'once');

% Load input dataset.
X = csvread(inputFile{:});

% Find out what dimension we want.
k = str2double(regexp(cmd,'.* -d.* (\d+)','tokens','once'));
% Validate the parameter.
if k > 0
    if k > size(X, 2)
        msg = [...
            '[Fatal] New dimensionality (%i) cannot be greater than'...
            'existing dimensionality (%i)!'...
            ];
        disp(sprintf(msg, k, size(X, 2)))
        return
    end
end

total_time = tic;
% Retrieve the dimensions of X.
[m, n] = size(X);

% Get the options for running PCA.
if strfind(cmd, '-s') > 0
    % The princomp function centers X by subtracting off column means, but 
    % the function doesn't rescale the columns of X. So we have to rescale
    % before princomp. If X is m-by-n with m > n, then compute only the 
    % first n columns.
    if (m <= n)
        [~, score] = princomp(zscore(X));
    else
        [~, score] = princomp(zscore(X), 'econ');
    end        
else
    % Performs principal components analysis on the dataset X. If X is 
    % m-by-n with m > n, then compute only the first n columns.
    if (m <= n)
        [~, score] = princomp(X);
    else
        [~, score] = princomp(X, 'econ');
    end        
end

% Reduced data dimension.
if k > 0
   score = score(:,1:k);
end

disp(sprintf('[INFO ]   total_time: %fs', toc(total_time)))

end

File: methods/weka/src/PCA.java
/**
 * @file PCA.java
 * @author Marcus Edel
 *
 * Principal Components Analysis with weka.
 */

import weka.core.Instances;
import weka.core.Utils;
import weka.core.converters.ConverterUtils.DataSource;
import weka.attributeSelection.AttributeSelection;
import weka.attributeSelection.PrincipalComponents;
import weka.attributeSelection.Ranker;

/**
 * This class use the weka libary to implement Principal Components Analysis.
 */
public class PCA {

  private static final String USAGE = String
      .format("This program performs principal components analysis on the given"
      + "dataset.\nIt will transform the data onto its principal components, "
      + "optionally performing\ndimensionality reduction by ignoring the "
      + "principal components with the\nsmallest eigenvalues.\n\n"
      + "Required options:\n"
      + "-i [string]     Input dataset to perform PCA on.\n\n"
      + "Options:\n\n"
      + "-d [int]    Desired dimensionality of output dataset. If -1,\n"
      + "            no dimensionality reduction is performed.\n"
      + "            Default value -1.\n"
      + "-s          If set, the data will be scaled before running\n"
      + "            PCA, such that the variance of each feature is 1.");

  public static void main(String args[]) {
    Timers timer = new Timers();
    try {
      // Get the data set path.
      String dataset = Utils.getOption('i', args);
      if (dataset.length() == 0)
        throw new IllegalArgumentException();

      timer.StartTimer("total_time");
      timer.StartTimer("loading_data");

      // Load input dataset.
      DataSource source = new DataSource(dataset);
      Instances data = source.getDataSet();

      timer.StopTimer("loading_data");

      // Find out what dimension we want.
      int k = 0;
      String dimension = Utils.getOption('d', args);
      if (dimension.length() == 0) {
        k = data.numAttributes();
      } else {
        k = Integer.parseInt(dimension);
        // Validate the parameter.
        if (k > data.numAttributes()) {
          System.out.printf("[Fatal] New dimensionality (%d) cannot be greater"
              + "than existing dimensionality (%d)!'\n", k, 
              data.numAttributes());
          
          System.exit(-1);
        }
      }

      // Performs a principal components analysis.
      PrincipalComponents pcaEvaluator = new PrincipalComponents();

      // Sets the amount of variance to account for when retaining principal 
      // components.
      pcaEvaluator.setVarianceCovered(1.0);
      // Sets maximum number of attributes to include in transformed attribute 
      // names.
      pcaEvaluator.setMaximumAttributeNames(-1);

      // Scaled X such that the variance of each feature is 1.
      boolean scale = Utils.getFlag('s', args);
      if (scale) {
        pcaEvaluator.setCenterData(true);
      } else {
        pcaEvaluator.setCenterData(false);
      }

      // Ranking the attributes.
      Ranker ranker = new Ranker();
      // Specify the number of attributes to select from the ranked list.
      ranker.setNumToSelect(k - 1);

      AttributeSelection selector = new AttributeSelection();
      selector.setSearch(ranker);
      selector.setEvaluator(pcaEvaluator);
      selector.SelectAttributes(data);

      // Transform data into eigenvector basis.
      Instances transformedData = selector.reduceDimensionality(data);

      timer.StopTimer("total_time");

      timer.PrintTimer("loading_data");
      timer.PrintTimer("total_time");

    } catch (IllegalArgumentException e) {
      System.err.println(USAGE);
    } catch (Exception e) {
      e.printStackTrace();
    }
  }
}
